src/test/java/com/google/research/bleth/services/DatabaseServiceTest.java

```diff
@@ -0,0 +1,264 @@
+        Board firstBoard = new Board(2, 2);
+        Board secondBoard = new Board(2, 2);
+        Mockito.when(firstAgent.moveTo()).thenReturn(oneOnOneCoordinate);
+        Mockito.when(secondAgent.moveTo()).thenReturn(zeroOnOneCoordinate);
+        firstBoard.placeAgent(firstAgent.moveTo(), firstAgent);
+        firstBoard.placeAgent(secondAgent.moveTo(), secondAgent);
+        secondBoard.placeAgent(secondAgent.moveTo(), secondAgent);
+        String firstBoardState = firstBoard.getState();
+        String secondBoardState = firstBoard.getState();
+
+        db.writeRealBoardState(firstSimulationId, roundOne, firstBoardState);
+
+        assertThrows(PreparedQuery.TooManyResultsException.class, () -> {
+            db.writeRealBoardState(firstSimulationId, roundOne, secondBoardState);
+        });
+    }
+
+    @Test
+    public void writeEstimatedBoardStateFromSameBoardsWithSameSimulationIdAndRound_ShouldThrowException() {
+
+        DatabaseService db = DatabaseService.getInstance();
+        Board board = new Board(2, 2);
+        Mockito.when(firstAgent.moveTo()).thenReturn(oneOnOneCoordinate);
+        Mockito.when(secondAgent.moveTo()).thenReturn(zeroOnOneCoordinate);
+        board.placeAgent(firstAgent.moveTo(), firstAgent);
+        board.placeAgent(secondAgent.moveTo(), secondAgent);
+        String boardState = board.getState();
+
+        db.writeEstimatedBoardState(firstSimulationId, roundOne, boardState);
+
+        assertThrows(PreparedQuery.TooManyResultsException.class, () -> {
+            db.writeEstimatedBoardState(firstSimulationId, roundOne, boardState);
+        });
+    }
+
+    @Test
+    public void writeEstimatedBoardStateFromDifferentBoardsWithSameSimulationIdAndRound_ShouldThrowException() {
+
+        DatabaseService db = DatabaseService.getInstance();
+        Board firstBoard = new Board(2, 2);
+        Board secondBoard = new Board(2, 2);
+        Mockito.when(firstAgent.moveTo()).thenReturn(oneOnOneCoordinate);
+        Mockito.when(secondAgent.moveTo()).thenReturn(zeroOnOneCoordinate);
+        firstBoard.placeAgent(firstAgent.moveTo(), firstAgent);
+        firstBoard.placeAgent(secondAgent.moveTo(), secondAgent);
+        secondBoard.placeAgent(secondAgent.moveTo(), secondAgent);
+        String firstBoardState = firstBoard.getState();
+        String secondBoardState = firstBoard.getState();
+
+        db.writeEstimatedBoardState(firstSimulationId, roundOne, firstBoardState);
+
+        assertThrows(PreparedQuery.TooManyResultsException.class, () -> {
+            db.writeEstimatedBoardState(firstSimulationId, roundOne, secondBoardState);
+        });
+    }
+
+    @After
+    public void tearDown() {
+        helper.tearDown();
+    }
+
+    @AfterAll
+    static void clearDatastore() {
+        DatabaseService db = DatabaseService.getInstance();
+        db.deleteAllSimulationRealBoardStates(firstSimulationId);
+        db.deleteAllSimulationEstimatedBoardStates(firstSimulationId);
```

Comment:
Isn't there any boardState with secondSimulationId?

Reply:
Correct, but I just figured out that this method is not needed at all - thanks to the helper and the tearDown() method (the helper provides a local datastore service which is designated for testing only):

https://cloud.google.com/appengine/docs/standard/java/tools/localunittesting